tests/unit/commands/test_completions.py

```diff
@@ -0,0 +1,143 @@
+    has_choice_options,
+    has_exact_options,
+    has_file_options,
+    has_unknown_required_param_options,
+)
+
+OptionType = namedtuple('OptionType', ['file', 'choice', 'unknown_required'])
+TestCase = Tuple[Action, OptionType]
+
+test_data: list[TestCase] = [
+    (Action('--flag', 'flag'), OptionType(file=False, choice=False, unknown_required=False)),
+    *(
+        (
+            Action('--flag', 'flag', metavar=metavar),
+            OptionType(file=True, choice=False, unknown_required=False),
+        )
+        for metavar in ('FILENAME', 'PATH')
+    ),
+    (
+        Action('--flag', dest='config_paths'),
+        OptionType(file=True, choice=False, unknown_required=False),
+    ),
+    (
+        Action('--flag', 'flag', metavar='OTHER'),
+        OptionType(file=False, choice=False, unknown_required=False),
+    ),
+    (
+        Action('--flag', 'flag', choices=['a', 'b']),
+        OptionType(file=False, choice=True, unknown_required=False),
+    ),
+    (
+        Action('--flag', 'flag', choices=['a', 'b'], type=str),
+        OptionType(file=False, choice=True, unknown_required=True),
+    ),
+    (
+        Action('--flag', 'flag', choices=None),
+        OptionType(file=False, choice=False, unknown_required=False),
+    ),
+    (
+        Action('--flag', 'flag', required=True),
+        OptionType(file=False, choice=False, unknown_required=True),
+    ),
+    *(
+        (
+            Action('--flag', 'flag', nargs=nargs),
+            OptionType(file=False, choice=False, unknown_required=True),
+        )
+        for nargs in ('+', '*')
+    ),
+    *(
+        (
+            Action('--flag', 'flag', metavar=metavar),
+            OptionType(file=False, choice=False, unknown_required=True),
+        )
+        for metavar in ('PATTERN', 'KEYS', 'N')
+    ),
+    *(
+        (
+            Action('--flag', 'flag', type=type, default=None),
+            OptionType(file=False, choice=False, unknown_required=True),
+        )
+        for type in (int, str)
+    ),
+    (
+        Action('--flag', 'flag', type=int, default=1),
+        OptionType(file=False, choice=False, unknown_required=False),
+    ),
+    (
+        Action('--flag', 'flag', type=str, required=True, metavar='PATH'),
+        OptionType(file=True, choice=False, unknown_required=True),
+    ),
+    (
+        Action('--flag', 'flag', type=str, required=True, metavar='PATH', default='/dev/null'),
+        OptionType(file=True, choice=False, unknown_required=True),
+    ),
+    (
+        Action('--flag', 'flag', type=str, required=False, metavar='PATH', default='/dev/null'),
+        OptionType(file=True, choice=False, unknown_required=False),
+    ),
+]
+
+
+@pytest.mark.parametrize('action, option_type', test_data)
+def test_has_file_options_detects_file_options(action: Action, option_type: OptionType):
+    assert (
+        has_file_options(action) == option_type.file
+    ), f'Action: {action} should be file={option_type.file}'
```

Comment:
IMO you might not need these messages. The default pytest assert introspection is pretty good these days and will produce a useful error message on its own. (Unless of course you've found otherwise...)

Reply:
sounds great, I have no experience with it, I just know assert doesn't produce nice messages in java and reached for an explicit message.

Reply:
Feel free to play with it and decide for yourself (or not!)...